sdk/core/azure_core_test/src/http/clients.rs

```diff
@@ -0,0 +1,75 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+use async_trait::async_trait;
+use azure_core::{HttpClient, Request, Response, Result};
+use futures::lock::Mutex;
+use std::{fmt, future::Future, pin::Pin};
+
+pub struct MockHttpClient<C>(Mutex<C>);
+
+impl<C> MockHttpClient<C>
+where
+    C: FnMut(&Request) -> Pin<Box<dyn Future<Output = Result<Response>> + Send + Sync + '_>>
+        + Send
+        + Sync,
+{
+    pub fn new(client: C) -> Self {
+        Self(Mutex::new(client))
+    }
+}
+
+impl<C> fmt::Debug for MockHttpClient<C> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str(stringify!("MockHttpClient"))
+    }
+}
+
+#[cfg_attr(target_arch = "wasm32", async_trait(?Send))]
+#[cfg_attr(not(target_arch = "wasm32"), async_trait)]
+impl<C> HttpClient for MockHttpClient<C>
+where
+    C: FnMut(&Request) -> Pin<Box<dyn Future<Output = Result<Response>> + Send + Sync + '_>>
+        + Send
+        + Sync,
```

Comment:
@analogrelay any other good ideas here? The DX this forces (see test below; we have to pin the future) is unwieldy but I couldn't figure out a better way without taking advantage of Rust 1.85 async captures.

Reply:
I cleaned it up a little using a `BoxedFuture`, though the `boxed()` "extension method" probably does more heavy lifting here. Still open to other suggestions, but this seems to be a fairly common solution I found - more than what inspired me originally.